lib/rgeo/cartesian/planar_graph.rb

```diff
@@ -0,0 +1,400 @@
+# frozen_string_literal: true
+
+module RGeo
+  module Cartesian
+    module Graphs
+      # HalfEdge represents an edge as 2 directed edges.
+      # One half-edge will have it's origin at edge.s, the other
+      # at edge.e. Both half-edges will be linked as each other's twins.
+      #
+      # HalfEdges also contain references to the next and prev half-edges,
+      # where next's origin is this half-edges destination. Prev's destination
+      # is this half-edge's origin.
+      class HalfEdge
+        include Comparable
+
+        # Creates 2 half edges from an edge.
+        # They will be assigned as each other's twins.
+        # The Half Edges will be returned in the order of points
+        # of the edge (start, end).
+        #
+        # @param edge [RGeo::Cartesian::Segment]
+        #
+        # @return [Array]
+        def self.from_edge(edge)
+          # TODO: should we raise an error if the edge is degenerate?
+          e1 = new(edge.s)
+          e2 = new(edge.e)
+
+          e1.twin = e2
+          e2.twin = e1
+          [e1, e2]
+        end
+
+        def initialize(origin)
+          @origin = origin
+          @twin = nil
+          @next = nil
+          @prev = nil
+        end
+        attr_reader :origin
+        attr_accessor :twin, :next, :prev
```

Comment:
Perhaps not the smartest to use `next` since that's a keyword. I would use `n` and `p` but `p` is also a keyword.

Reply:
Next is okay since it is a keyword and we are most likely to call it in a context that has no ambiguity. Since p is a method it would be worse. 

A solution though if you prefer : `next_hedge`

Reply:
That's a good point. I'll likely leave it as `next` then.